lib/std/dns/rdata.zig

```diff
@@ -0,0 +1,236 @@
+        .PTR => DNSRData{ .PTR = try pkt.deserializeName(&deserializer) },
+        .MX => blk: {
+            break :blk DNSRData{
+                .MX = MXData{
+                    .preference = try deserializer.deserialize(u16),
+                    .exchange = try pkt.deserializeName(&deserializer),
+                },
+            };
+        },
+        .MD => DNSRData{ .MD = try pkt.deserializeName(&deserializer) },
+        .MF => DNSRData{ .MF = try pkt.deserializeName(&deserializer) },
+
+        .SOA => blk: {
+            var mname = try pkt.deserializeName(&deserializer);
+            var rname = try pkt.deserializeName(&deserializer);
+            var serial = try deserializer.deserialize(u32);
+            var refresh = try deserializer.deserialize(u32);
+            var retry = try deserializer.deserialize(u32);
+            var expire = try deserializer.deserialize(u32);
+            var minimum = try deserializer.deserialize(u32);
+
+            break :blk DNSRData{
+                .SOA = SOAData{
+                    .mname = mname,
+                    .rname = rname,
+                    .serial = serial,
+                    .refresh = refresh,
+                    .retry = retry,
+                    .expire = expire,
+                    .minimum = minimum,
+                },
+            };
+        },
+
+        else => blk: {
+            return error.InvalidRData;
+        },
+    };
+
+    return rdata;
+}
+
+/// Serialize a given DNSRData into []u8
+pub fn serializeRData(
+    rdata: DNSRData,
+    serializer: var,
+) !void {
+    switch (rdata) {
+        .A => |addr| try serializer.serialize(addr.in.addr),
+        .AAAA => |addr| try serializer.serialize(addr.in6.addr),
+
+        .NS, .MD, .MF, .MB, .MG, .MR, .CNAME, .PTR => |name| try serializer.serialize(name),
+
+        .SOA => |soa_data| blk: {
+            try serializer.serialize(soa_data.mname);
+            try serializer.serialize(soa_data.rname);
+
+            try serializer.serialize(soa_data.serial);
+            try serializer.serialize(soa_data.refresh);
+            try serializer.serialize(soa_data.retry);
+            try serializer.serialize(soa_data.expire);
+            try serializer.serialize(soa_data.minimum);
+        },
+
+        .MX => |mxdata| blk: {
+            try serializer.serialize(mxdata.preference);
+            try serializer.serialize(mxdata.exchange);
+        },
+
+        else => return error.NotImplemented,
+    }
+}
+
+fn printName(
+    stream: *std.io.OutStream(OutError),
+    name: dns.DNSName,
+) !void {
+    // This doesn't use the name since we can just write to the stream.
+    for (name.labels) |label| {
+        try stream.print("{}.", .{label});
+    }
+}
+
+/// Prettify a given DNSRData union variable. Returns a string with e.g
+/// a string representation of an ipv4 address, or the human-readable version
+/// of a DNSName.
+pub fn prettyRData(allocator: *std.mem.Allocator, rdata: DNSRData) ![]const u8 {
```

Comment:
I think this (and printName) would work better as a custom formatter. Then it could attach itself to any existing formatter / stream / buffer, and this body would be replaced with `allocPrint(allocator, "{}", .{rdata})`